src/utils/change_resolver.ts

```diff
@@ -0,0 +1,58 @@
+import { Wallet } from "ethers";
+import { providers } from "ethers";
+import { namehash } from "ethers/utils";
+import { EnsRegistryFactory } from "../../ethers/EnsRegistryFactory";
+import { PublicResolverFactory } from "../../ethers/PublicResolverFactory";
+import { getSubdomains } from "../utils/getSubDomains";
+
+const { JsonRpcProvider } = providers;
+
+export interface ChangeResolverParams {
+  rootNode: string,
+  privateKey: string,
+  rpcUrl: string,
+  registryAddr: string,
+  resolverAddr: string,
+  newResolverAddr: string
+}
+
+const owners: Record<string, string> = {};
+
+/**
+ * @description - Updates resolver on all subnodes of `parentNode`
+ * @param rootNode 
+ * @param resolverAddr 
+ */
+export async function changeResolver(
+  { rootNode, privateKey, rpcUrl, registryAddr, resolverAddr, newResolverAddr }:
+    ChangeResolverParams
+) {
+  const rootOwner = new Wallet(privateKey).address;
+
+  const provider = new JsonRpcProvider(rpcUrl);
+  const wallet = new Wallet(privateKey, provider);
+  const ensRegistry = EnsRegistryFactory.connect(registryAddr, wallet);
+  const ensResolver = PublicResolverFactory.connect(resolverAddr, wallet);
+
+  const changeDomainResolver = async (domainRoot: string) => {
+    await ensRegistry.setResolver(namehash(domainRoot), newResolverAddr);
+    const subnodes = await getSubdomains({ domain: domainRoot, ensResolver, ensRegistry, mode: "ALL" });
+    for (const node of subnodes) {
+      const owner = await ensRegistry.functions.owner(namehash(node));
+      owners[node] = owner;
+
+      await ensRegistry.setOwner(namehash(node), rootOwner);
+      await ensRegistry.setResolver(namehash(node), newResolverAddr);
+    }
+  };
+
+  const restoreOwners = async () => {
+    for (const node in owners) {
+      await ensRegistry.setOwner(namehash(node), owners[node]);
```

Comment:
I'm surprised this works as the rootOwner might not own intermediary nodes anymore. See this test I made: https://github.com/energywebfoundation/iam-contracts/blob/af60a5ad99cc763a3ccf14b668ed675e51d8227f/test/RoleDefinitionResolver.Test.ts#L104
Maybe I'm missing something though 🤔

Reply:
`changeDomainResolver` traverses all nodes under parent and changes their owner to on `rootOwner` in order to be able to change resolver

Reply:
I could be wrong but I think this test should pass, but it fails. Please try it out.

```
export const utilsTests = () => {
  const newResolverAddr = emptyAddress;

  test('resolver can be changed', async () => {
    const rootNode = `${root}`; // Changed to just root...

    const params = {
      rootNode,
      privateKey: rootOwner.privateKey,
      registryAddr: ensRegistry.address,
      resolverAddr: ensResolver.address,
      rpcUrl,
      newResolverAddr
    };

    // Print out current owners (all same address)
    const actualRootOwner = await ensRegistry.owner(namehash(root));
    console.info("actual root owner", actualRootOwner);
    const org1owner = await ensRegistry.owner(namehash(rootNode));
    console.info("org1 owner", org1owner);
    const domains_1 = await getSubdomains({ domain: rootNode, ensResolver, ensRegistry, mode: 'ALL' });
    await Promise.all(domains_1.map(async domain => {
      const owner = await ensRegistry.owner(namehash(domain));
      console.info(`${domain} owner`, owner);
    }));

    // Set to new address (so orginal address owns the root, doesn't own the org or the sub org)
    await ensRegistry.setOwner(namehash('org1-1.org1.root'), '0xE45Ad1e7522288588dA6829A9ea6A09e92FCDe14');
    await ensRegistry.setOwner(namehash('org1.root'), '0xE45Ad1e7522288588dA6829A9ea6A09e92FCDe14');

    await changeResolver(params);

    const domains = await getSubdomains({ domain: rootNode, ensResolver, ensRegistry, mode: 'ALL' });
    const resolvers = await Promise.all(domains.map(async (domain) => await ensRegistry.resolver(namehash(domain))));

    expect(
      resolvers
        .every((addr) => addr === newResolverAddr))
      .toBe(true);
  });
};
```

Reply:
Thanks. You have found serious bug. I should have changed the owner using `setSubnodeOwner` and not `setOwner`